File: android-client/app/src/main/java/dev/sever/vpn/Wire.java
package dev.sever.vpn;

import java.io.*;
import java.security.SecureRandom;
import java.util.*;

public final class Wire {
    public static final int AUTH=1, CONFIG=2, PACKET=3, PING=4, PONG=5, CLOSE=6, MTU=1280;
    public static final int MAX_FRAME=4615;
    private static final SecureRandom RANDOM = new SecureRandom();
    public static final class Frame {
        public final int kind; public final byte[] data;
        Frame(int kind, byte[] data) { this.kind=kind; this.data=data; }
    }
    public static byte[] encode(int kind, byte[] data) throws IOException {
        ByteArrayOutputStream out=new ByteArrayOutputStream(); write(out,kind,data); return out.toByteArray();
    }
    public static void write(OutputStream stream, int kind, byte[] data) throws IOException {
        if (kind < 1 || kind > 6 || data.length > 4096) throw new IOException("Invalid frame");
        // Data packets are intentionally unpadded in VISION performance mode. Control frames keep
        // small random padding for wire compatibility without wasting up to 511 bytes per IP packet.
        int padding = kind==PACKET ? 0 : RANDOM.nextInt(96);
        DataOutputStream out = new DataOutputStream(stream);
        out.writeByte(1); out.writeByte(kind); out.writeShort(data.length); out.writeShort(padding); out.writeShort(0);
        out.write(data);
        if(padding>0){byte[] pad=new byte[padding];RANDOM.nextBytes(pad);out.write(pad);}
        out.flush();
    }
    public static Frame read(InputStream stream) throws IOException {
        DataInputStream in = new DataInputStream(stream);
        int version=in.readUnsignedByte(), kind=in.readUnsignedByte();
        int length=in.readUnsignedShort(), padding=in.readUnsignedShort(), reserved=in.readUnsignedShort();
        if(version!=1 || kind<1 || kind>6 || length>4096 || padding>511 || reserved!=0) throw new IOException("Invalid frame header");
        byte[] payload=new byte[length], pad=new byte[padding]; in.readFully(payload); in.readFully(pad); return new Frame(kind,payload);
    }
    public static Frame decode(byte[] bytes) throws IOException {
        List<Frame> frames=decodeMany(bytes); if(frames.size()!=1)throw new IOException("One frame required"); return frames.get(0);
    }
    public static List<Frame> decodeMany(byte[] bytes) throws IOException {
        if(bytes.length<8 || bytes.length>65536)throw new IOException("Invalid message size");
        ByteArrayInputStream input=new ByteArrayInputStream(bytes); ArrayList<Frame> frames=new ArrayList<>();
        while(input.available()>0){ if(input.available()<8)throw new IOException("Truncated frame"); frames.add(read(input)); }
        return frames;
    }
    public static byte[] address(String value) throws IOException {
        String[] parts=value.split("\\.",-1); if(parts.length!=4)throw new IOException("IPv4 required"); byte[] bytes=new byte[4];
        for(int i=0;i<4;i++){if(!parts[i].matches("[0-9]{1,3}"))throw new IOException("Invalid IPv4");int n=Integer.parseInt(parts[i]);if(n>255)throw new IOException("Invalid IPv4");bytes[i]=(byte)n;}return bytes;
    }
    public static void validatePacket(byte[] p, byte[] address, boolean source) throws IOException {
        if(p.length<20 || p.length>MTU || (p[0]&255)>>4!=4)throw new IOException("Invalid IPv4 packet");
        int ihl=(p[0]&15)*4,total=((p[2]&255)<<8)|(p[3]&255);if(ihl<20||ihl>p.length||total!=p.length)throw new IOException("Invalid packet length");
        if(!Arrays.equals(address,Arrays.copyOfRange(p,source?12:16,source?16:20)))throw new IOException("Packet address mismatch");
    }
}

File: ios-client/Shared/Wire.swift
import Foundation
import Security

enum Wire {
    static let auth: UInt8 = 1, config: UInt8 = 2, packet: UInt8 = 3, ping: UInt8 = 4, pong: UInt8 = 5
    static func frame(_ kind: UInt8, _ payload: Data) throws -> Data {
        guard (1...6).contains(kind), payload.count <= 4096 else { throw SeverError.invalidFrame }
        let padding = Int.random(in: 0...511)
        var out = Data([1, kind, UInt8(payload.count >> 8), UInt8(payload.count & 255), UInt8(padding >> 8), UInt8(padding & 255), 0, 0])
        out.append(payload)
        var bytes = [UInt8](repeating: 0, count: padding)
        if padding > 0 && SecRandomCopyBytes(kSecRandomDefault, padding, &bytes) != errSecSuccess { throw SeverError.invalidFrame }
        out.append(contentsOf: bytes)
        return out
    }
    static func header(_ data: Data) throws -> (UInt8, Int, Int) {
        let b = [UInt8](data)
        guard b.count == 8, b[0] == 1, (1...6).contains(b[1]), b[6] == 0, b[7] == 0 else { throw SeverError.invalidFrame }
        let n = Int(b[2]) * 256 + Int(b[3]), pad = Int(b[4]) * 256 + Int(b[5])
        guard n <= 4096, pad <= 511 else { throw SeverError.invalidFrame }
        return (b[1], n, pad)
    }
    static func decode(_ data: Data) throws -> (UInt8, Data) {
        guard (8...4615).contains(data.count) else { throw SeverError.invalidFrame }
        let (kind, length, padding) = try header(Data(data.prefix(8)))
        guard data.count == 8 + length + padding else { throw SeverError.invalidFrame }
        return (kind, Data(data.dropFirst(8).prefix(length)))
    }
    static func address(_ text: String) throws -> [UInt8] {
        let pieces = text.split(separator: ".", omittingEmptySubsequences: false)
        let bytes = pieces.compactMap { UInt8($0) }
        guard pieces.count == 4, bytes.count == 4 else { throw SeverError.invalidFrame }
        return bytes
    }
    static func validate(_ packet: Data, address: [UInt8], source: Bool) throws {
        let b = [UInt8](packet)
        guard b.count >= 20, b.count <= 1280, b[0] >> 4 == 4 else { throw SeverError.invalidFrame }
        let ihl = Int(b[0] & 15) * 4
        guard ihl >= 20, ihl <= b.count, Int(b[2]) * 256 + Int(b[3]) == b.count,
              Array(b[(source ? 12 : 16)..<(source ? 16 : 20)]) == address else { throw SeverError.invalidFrame }
    }
}
